fix: Add input validation to EditProfileFragment

haroldadmin opened the pull request:
This commit adds input validation to the EditProfileFragment.

### Description
An extra file called States.kt has been created to represent possible
error states (there is only one right now). More errors can be added to
this class as the application grows and the number of required fields in
a profile increases.

Fixes #110 

### Type of Change:
**Delete irrelevant options.**

- Code
- Quality Assurance

**Code/Quality Assurance Only**
- Bug fix (non-breaking change which fixes an issue)

### How Has This Been Tested?
Tested on my phone. Attached a screenshot to show implementation of this feature.
![image 2018-12-11 11 45 19](https://user-images.githubusercontent.com/24315306/49781859-3e4d8b80-fd3a-11e8-89fe-fcda7628bea8.jpg)

### Checklist:
**Delete irrelevant options.**

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings

isabelcosta commented on app/src/main/java/org/systers/mentorship/utils/States.kt:
```diff
@@ -0,0 +1,11 @@
+ * grows in complexity.
+ */
+sealed class EditProfileFragmentErrorStates {
+    object EmptyNameError: EditProfileFragmentErrorStates()
```
Codacy is recommending to add KotlinDocs for this object [(link to Codacy)](https://app.codacy.com/app/m-murad/mentorship-android/pullRequest?prid=2718301)

haroldadmin replied:
...and it's done 😊

isabelcosta commented on app/src/main/java/org/systers/mentorship/view/fragments/EditProfileFragment.kt:
```diff
@@ -97,4 +105,20 @@ class EditProfileFragment: DialogFragment() {
+     * be empty.
+     */
+    private fun validateProfileInput(name: String?): Array<EditProfileFragmentErrorStates> {
+        var errors = arrayOf<EditProfileFragmentErrorStates>()
```
I like this modular implementation :)

haroldadmin replied:
Thank you! I wanted this feature to extensible, so I opted for this approach.

isabelcosta commented on app/src/main/java/org/systers/mentorship/view/fragments/EditProfileFragment.kt:
```diff
@@ -97,4 +105,20 @@ class EditProfileFragment: DialogFragment() {
     }
+
+    /**
+     * A function to perform input validation
```
This is too much broken into lines, you have 8 lines that could be better distributed right?

isabelcosta replied:
Should we do here length validations how it is on the backend? @m-murad

haroldadmin replied:
@isabelcosta Thanks for the suggestion. I have made the this change. The comments now have longer lines. I'm also adding length validations, just to be safe (yay for the modular system!)

isabelcosta reviewed (changes requested):
Good job @haroldadmin :)
Just left some minor change requests

haroldadmin commented:
@isabelcosta I have made the requested changes. Please review when possible!

isabelcosta commented on app/src/main/java/org/systers/mentorship/view/fragments/EditProfileFragment.kt:
```diff
@@ -97,4 +111,19 @@ class EditProfileFragment: DialogFragment() {
+    private fun validateProfileInput(name: String?): Array<EditProfileFragmentErrorStates> {
+        var errors = arrayOf<EditProfileFragmentErrorStates>()
+        if (name.isNullOrEmpty()) errors += EditProfileFragmentErrorStates.EmptyNameError
+        if (name?.length ?: 0 < 2) errors += EditProfileFragmentErrorStates.NameTooShortError
```
should this be on integer resources instead of hardcoded in this code? cc @m-murad

haroldadmin replied:
Okay, I'll add this as an Integer resource.

isabelcosta commented on app/src/main/java/org/systers/mentorship/view/fragments/EditProfileFragment.kt:
```diff
@@ -82,11 +83,24 @@ class EditProfileFragment: DialogFragment() {
+                    is EditProfileFragmentErrorStates.EmptyNameError ->
+                        getString(R.string.error_empty_name)
+                    is EditProfileFragmentErrorStates.NameTooShortError ->
+                        "Name should have a minimum of 2 characters"
```
move this to string resources please

haroldadmin replied:
Okay, moved this to a string resource.

isabelcosta commented on app/src/main/java/org/systers/mentorship/view/fragments/EditProfileFragment.kt:
```diff
@@ -82,11 +83,24 @@ class EditProfileFragment: DialogFragment() {
+                    is EditProfileFragmentErrorStates.NameTooShortError ->
+                        "Name should have a minimum of 2 characters"
+                    is EditProfileFragmentErrorStates.NameTooLongError ->
+                        "Name should have a maximum of 30 characters"
```
move this to string resources please

haroldadmin commented:
@isabelcosta @m-murad I have added the requested changes.

isabelcosta commented on app/src/main/java/org/systers/mentorship/view/fragments/EditProfileFragment.kt:
```diff
@@ -82,11 +83,24 @@ class EditProfileFragment: DialogFragment() {
+                    is EditProfileFragmentErrorStates.EmptyNameError ->
+                        getString(R.string.error_empty_name)
+                    is EditProfileFragmentErrorStates.NameTooShortError ->
+                        context.getString(R.string.error_name_too_short)
```
How can you getString without calling from context in line 91, but you call it here? @haroldadmin

haroldadmin replied:
I'm sorry, this probably happened during the extraction of the string resource. It succesfully builds because the fragment class itself also has a method called `getString`, and so it slipped past me. I'll fix this right now.

isabelcosta commented on app/src/main/res/values/strings.xml:
```diff
@@ -99,4 +99,6 @@
     <string name="notes_empty_error">Notes cannot be empty</string>
     <string name="available_to_be_mentor_mentee">Available to be a:</string>
     <string name="settings">Settings</string>
+    <string name="error_name_too_short">Name should have a minimum of 2 characters</string>
```
Is it too much to ask for this string resource to be a formatted string, that accepts an integer in the place of the size of the name (same for the line below)

This is because, if we need to change these constraints, we'll only need to change in the integer resources.

What do you think?

haroldadmin replied:
I think that's a good idea, it's just gonna be one less thing to update in the future in case we change character count requirements. I'll have to look into how to use formatted strings in XML though. I'll push an update soon.

isabelcosta commented on app/src/main/res/values/constants.xml:
```diff
@@ -0,0 +1,5 @@
+<resources>
+    <integer name="min_name_length">2</integer>
+    <integer name="max_name_length">30</integer>
+</resources>
```
a blank line at the end of this file, please :)

haroldadmin replied:
Aha! Missing blank lines at the end lint check finds another victim! 😂 
I'll add this and update the PR soon.

isabelcosta reviewed (changes requested):
I requested 2 changes. The last one I think its important. After you do the changes I'm ok with merging this!

haroldadmin commented:
Whoops, I made a mistake of accidentally force pushing without any changes, and it closed the PR automatically.

haroldadmin commented:
All done now, hopefully. Please review when you have time @isabelcosta

isabelcosta commented:
@haroldadmin no problem. I think I did that once during GSoC :P I'm glad you reopened this instead of creating another PR

isabelcosta commented on app/src/main/java/org/systers/mentorship/utils/States.kt:
```diff
@@ -0,0 +1,32 @@
+    /**
+     * An object to represent the error case when the
+     * entered name of the user is too short. Currently
+     * the mentorship-backend has this limit set to two
```
no need to mention character limit here

haroldadmin replied:
I've removed that sentence now.

isabelcosta commented on app/src/main/java/org/systers/mentorship/utils/States.kt:
```diff
@@ -0,0 +1,32 @@
+    /**
+     * An object to represent the error case when the
+     * entered name of the user is too long. Currently
+     * the mentoship-backend has this limit set to 30
```
no need to mention character limit here

haroldadmin replied:
I've removed that sentence now.

isabelcosta reviewed (changes requested):
Only other 2 changes and then I can approve this 🎉

haroldadmin committed:
fix: Add input validation to EditProfileFragment

Fixes Issue #110. This commit adds input validation to the
EditProfileFragment. An extra file called States.kt has been created to represent possible
error states. More errors can be added to
this class as the application grows and the number of required fields in
a profile increases.

isabelcosta reviewed (approved):
Good job @haroldadmin :)
